pkg/metrics/metrics.go

```diff
@@ -0,0 +1,190 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package metrics contains global structures related to metrics collection
+// cert-manager exposes the following metrics:
+// certificate_requests{name, namespace, issuer_name, issuer_scope, issuer_type, operation, result}
+// certificate_expiry_time_seconds{name, namespace}
+package metrics
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/gorilla/mux"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	corelisters "k8s.io/client-go/listers/core/v1"
+
+	"github.com/jetstack/cert-manager/pkg/apis/certmanager/v1alpha1"
+	"github.com/jetstack/cert-manager/pkg/issuer"
+	"github.com/jetstack/cert-manager/pkg/util/kube"
+)
+
+const (
+	// Namespace is the namespace for cert-manager metric names
+	Namespace                              = "certmanager"
+	prometheusMetricsServerAddress         = "0.0.0.0:9402"
+	prometheusMetricsServerShutdownTimeout = 5 * time.Second
+	prometheusMetricsServerReadTimeout     = 8 * time.Second
+	prometheusMetricsServerWriteTimeout    = 8 * time.Second
+	prometheusMetricsServerMaxHeaderBytes  = 1 << 20
+
+	operationIssue = "issue"
+	operationRenew = "renew"
+	resultSuccess  = "success"
+	resultFailure  = "failure"
+)
+
+type Metrics struct {
+	http.Server
+
+	// TODO (@dippynark): switch this to use an interface to make it testable
+	registry                     *prometheus.Registry
+	CertificateRequests          *prometheus.CounterVec
+	CertificateExpiryTimeSeconds *prometheus.GaugeVec
+}
+
+func New() *Metrics {
+
+	router := mux.NewRouter()
+
+	// Create server and register prometheus metrics handler
+	s := &Metrics{
+		Server: http.Server{
+			Addr:           prometheusMetricsServerAddress,
+			ReadTimeout:    prometheusMetricsServerReadTimeout,
+			WriteTimeout:   prometheusMetricsServerWriteTimeout,
+			MaxHeaderBytes: prometheusMetricsServerMaxHeaderBytes,
+			Handler:        router,
+		},
+		registry: prometheus.NewRegistry(),
+		CertificateRequests: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Namespace: Namespace,
+				Name:      "certificate_requests",
+				Help:      "Number of certificate requests",
+			},
+			[]string{"name", "namespace", "issuer_name", "issuer_kind", "issuer_type", "operation", "result"},
+		),
+		CertificateExpiryTimeSeconds: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Namespace: Namespace,
+				Name:      "certificate_expiry_time_seconds",
+				Help:      "Number of seconds after January 1, 1970 UTC that the certificate will expire",
```

Comment:
`Number of seconds after January 1, 1970 UTC that the certificate will expire` so this is a unix timestamp in seconds of the expiry date, right? (as opposed to # of seconds until expiry). Makes sense (I think 🙄)

Reply:
I've changed this to `certificate_expiration_seconds` - `The date after which the certificate expires. Expressed as a Unix Epoch Time.`, which I think is clearer